src/alpaka/AlpakaCore/alpakaWorkDivHelper.h

```diff
@@ -29,50 +29,268 @@ namespace cms {
+        // Take into account the block index in grid.
+        const uint32_t blockIdxInGrid(alpaka::idx::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[dimIndex]);
+        const uint32_t blockDimension(alpaka::workdiv::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[dimIndex]);
+
+        // Shift to get global indices in grid (instead of local to the block)
+        elementIdxShift[dimIndex] += blockIdxInGrid * blockDimension;
+      }
+
+      // Return element indexes, shifted by elementIdxShift.
+      return element_index_range_in_block(acc, elementIdxShift);
+    }
+
+    /*
+     * Computes the range of the elements indexes in grid.
+     * Truncated by the max number of elements of interest.
+     */
+    template <typename T_Acc, typename T_Dim>
+    ALPAKA_FN_ACC std::pair<Vec<T_Dim>, Vec<T_Dim>> element_index_range_in_grid_truncated(
+        const T_Acc& acc, const Vec<T_Dim>& maxNumberOfElements, Vec<T_Dim>& elementIdxShift) {
+      // Check dimension
+      static_assert(alpaka::dim::Dim<T_Acc>::value == T_Dim::value,
+                    "Accelerator and maxNumberOfElements need to have same dimension.");
+      auto&& [firstElementIdxGlobalVec, endElementIdxGlobalVec] = element_index_range_in_grid(acc, elementIdxShift);
+
+      // Truncate
       for (typename T_Dim::value_type dimIndex(0); dimIndex < T_Dim::value; ++dimIndex) {
         endElementIdxGlobalVec[dimIndex] = std::min(endElementIdxGlobalVec[dimIndex], maxNumberOfElements[dimIndex]);
       }
 
+      // Return element indexes, shifted by elementIdxShift, and truncated by maxNumberOfElements.
       return {firstElementIdxGlobalVec, endElementIdxGlobalVec};
     }
 
+    /*
+     * Computes the range of the element(s) index(es) in grid.
+     * Truncated by the max number of elements of interest.
+     */
+    template <typename T_Acc, typename T_Dim>
+    ALPAKA_FN_ACC std::pair<Vec<T_Dim>, Vec<T_Dim>> element_index_range_in_grid_truncated(
+        const T_Acc& acc, const Vec<T_Dim>& maxNumberOfElements) {
+      Vec<T_Dim> elementIdxShift = Vec<T_Dim>::zeros();
+      return element_index_range_in_grid_truncated(acc, maxNumberOfElements, elementIdxShift);
+    }
+
+    /*********************************************
+     *     1D HELPERS, LOOP ON ALL CPU ELEMENTS
+     ********************************************/
+
+    /*
+     * Loop on all (CPU) elements.
+     * Elements loop makes sense in CPU case only. In GPU case, elementIdx = firstElementIdx = threadIdx + shift.
+     * Indexes are local to the BLOCK.
+     */
+    template <typename T_Acc, typename Func>
+    ALPAKA_FN_ACC void for_each_element_in_thread_1D_index_in_block(const T_Acc& acc,
+                                                                    const uint32_t maxNumberOfElements,
+                                                                    const uint32_t elementIdxShift,
+                                                                    const Func& func) {
```

Comment:
I was a little bit surprised that `const Func&` works also for mutating lambdas. I guess it works because the lambdas here capture by reference, and the language allows (too easly) a const pointer to non-const object.

Reply:
Yes will change it to just a pass by value, not as if the function object was heavy anyway

Reply:
Or universal reference `Func&& func`? I suppose that would be most agnostic to the details (in principle the function could get a lambda/functor with heavy-to-copy member data).